Show skipped resources

sinabakh opened the pull request:
Fixes #84 

Here's a sample output with no flag:
```
  NAME           MONTHLY QTY  UNIT  PRICE  HOURLY COST  MONTHLY COST  

  OVERALL TOTAL                                 0.0000        0.0000  

3 out of 3 resources were skipped, 1 are free, and 2 aren't supported in Infracost yet(https://www.infracost.io/docs/supported_resources), re-run with --show-skipped to see the list.
```

And with `--show-skipped` flag:
```
  NAME           MONTHLY QTY  UNIT  PRICE  HOURLY COST  MONTHLY COST  

  OVERALL TOTAL                                 0.0000        0.0000  

Skipped aws_proxy_configuration.my_proxy_configuration | This resource is free.
Skipped aws_dynamodb_table.my_dynamodb_table | This resource is not currently supported.
Skipped aws_db_instance.my_db_instance | This resource is not currently supported.
3 out of 3 resources were skipped, 1 are free, and 2 aren't supported in Infracost yet(https://www.infracost.io/docs/supported_resources), re-run with --show-skipped to see the list.
```

sinabakh commented:
@alikhajeh1 @aliscott It lacks tests and some final touches. But we can discuss this approach before doing them.
This seemed the most flexible and clean way of doing it to me.

alikhajeh1 commented:
Thank you :) I like where this is going! I thought more about it and I'm changing my mind on seeing free resources as explained below.

The output could look like this:
```
2 out of 26 resources couldn't be estimated as Infracost doesn't support them yet (https://www.infracost.io/docs/supported_resources), re-run with --show-skipped to see the list.

We're continually adding new resources, please create an issue if you'd like us to prioritize your list.
```

Re-running with `--show-skipped` could add this to the top of the above output (a sorted list showing the number of resources per unsupported resource type):
```
10 x aws_sns_platform_application
8 x aws_workspaces_directory
3 x aws_elastic_beanstalk_application
```

Stepping back, here's what I want as a user:
I want to run infracost, understand the output, and see a summary of why the estimate is very different to my last month's bill. If I want to dig deeper, I should be able to see details, but there is no need to show me my resource names, just the resource type and how many of them I'm using is fine. I could create an issue with that list to ask for infracost to support them. I highly prefer not to read docs. 

Long term, hopefully this feature can become mostly redundant as we support more resources.

Let's look at the different cases for resources:
1. There are resources like aws_ecs_capacity_provider that don't have a direct cost, they can include configs. Some of these configs might influence costs but that's already handled in the resource that uses the "config" resource. Infracost should not show these as it's just noise.

2. There are resources that infracost partially supports, e.g. aws_ecs_service.launch_type==FARGATE is supported but launch_type==EC2 is not. If a terraform project uses the non-supported resource config, infracost should count that as a non-supported resource. As a maintainer, I can see the `SkipMessage` to learn about what the details are.

3. There are resources that infracost doesn't support yet. That should obviously be counted as a non-supported resource.

For resource components:
There are "obviously free" resource components, e.g. ecs_task_definition.name. There are suspicious resource components like ecs_task_definition.docker_volume_configuration that they don't cost anything. As long as I trust that infracost is covering all components in the resource, then there is no need to see these in the output.

Finally, for infracost docs:
Maybe just copying the supported resource list from the resource registry is good enough and we can add a comment or json annotation with notes. That means we don't need to ask contributors to send a docs PR, one less step for them.

sinabakh commented:
So the outputs are like this:

Without the flag:
```
  NAME           MONTHLY QTY  UNIT  PRICE  HOURLY COST  MONTHLY COST  

  OVERALL TOTAL                                 0.0000        0.0000  

7 out of 7 resources couldn't be estimated as Infracost doesn't support them yet (https://www.infracost.io/docs/supported_resources), re-run with --show-skipped to see the list.
We're continually adding new resources, please create an issue if you'd like us to prioritize your list.
```

With `--show-skipped` flag:
```
  NAME           MONTHLY QTY  UNIT  PRICE  HOURLY COST  MONTHLY COST  

  OVERALL TOTAL                                 0.0000        0.0000  

7 out of 7 resources couldn't be estimated as Infracost doesn't support them yet (https://www.infracost.io/docs/supported_resources).
1 x aws_nat_gateway 
1 x aws_dynamodb_table 
2 x aws_ebs_volume 
1 x aws_eip 
1 x aws_instance 
1 x aws_lambda_function 
We're continually adding new resources, please create an issue if you'd like us to prioritize your list.
```

sinabakh commented:
I just can't figure out why go-lint fails, I think sth is wrong with the ci/cd pipeline.

aliscott commented:
> I just can't figure out why go-lint fails, I think sth is wrong with the ci/cd pipeline.

CI output is messed up for some reason but I do see this `##[error]unreachable: unreachable code (govet)`.

Running this branch locally gives me this lint error:
```
pkg/schema/resource.go:86:2: unreachable: unreachable code (govet)
	return len(r.CostComponents) == 0
```

sinabakh commented:
@aliscott Yeah I left a FIXME comment there, the output was wrong because of #121 but since we don't show free resources anymore I will fix it now.

alikhajeh1 commented on internal/providers/terraform/provider.go:
```diff
@@ -102,3 +102,40 @@ func generatePlanJSON(dir string, path string) ([]byte, error) {
+			fmt.Printf("%d x %s \n", count, rType)
+		}
+	}
+	fmt.Println("We're continually adding new resources, please create an issue if you'd like us to prioritize your list.")
```
This seems to look better when it's added in the `message` on line 128.

alikhajeh1 commented:
@sinabakh nice one!

It would be great if we can ensure that infracost produces valid json with `--output json` (I'll add a CI check soon). Can you possibly add a `warnings: ["message here"]` to the json output? The idea is that for now, it's just one long string in a warnings array, but we can split it out into other warning fields in the future. The string warning can include the resource type list if `--show-skipped` is used.

sinabakh commented:
There are some major changes with a few breaking changes:
* Moved the skipped resource message builder to the outputs
* Moved the logic of counting and aggregating the skipped resource to a method in the provider
* BREAKING: The json output used to be an array, I changed it to an object.

So here is a sample output for the table:
```
  NAME           MONTHLY QTY  UNIT  PRICE  HOURLY COST  MONTHLY COST  

  OVERALL TOTAL                                 0.0000        0.0000  

7 out of 7 resources couldn't be estimated as Infracost doesn't support them yet (https://www.infracost.io/docs/supported_resources).
We're continually adding new resources, please create an issue if you'd like us to prioritize your list.
1 x aws_dynamodb_table
2 x aws_ebs_volume
1 x aws_eip
1 x aws_instance
1 x aws_lambda_function
1 x aws_nat_gateway
```

And a sample output for the json:
```json
{"resources":[],"warnings":["\n7 out of 7 resources couldn't be estimated as Infracost doesn't support them yet (https://www.infracost.io/docs/supported_resources).\nWe're continually adding new resources, please create an issue if you'd like us to prioritize your list.\n1 x aws_eip\n1 x aws_instance\n1 x aws_lambda_function\n1 x aws_nat_gateway\n1 x aws_dynamodb_table\n2 x aws_ebs_volume\n"]}
```

aliscott commented on pkg/output/json.go:
```diff
@@ -52,12 +59,41 @@ func newResourceJSON(r *schema.Resource) resourceJSON {
+	if unSupportedCount == 0 {
+		return nil
+	}
+	message := fmt.Sprintf("\n%d out of %d resources couldn't be estimated as Infracost doesn't support them yet (https://www.infracost.io/docs/supported_resources)", unSupportedCount, len(resources))
```
Do we need a \n at the start here?

sinabakh replied:
In the table output it becomes hard to distinguish the aggrgated prices (last line of output) from this output but you are right I will remove it for now.

sinabakh replied:
Done

aliscott commented on pkg/output/json.go:
```diff
@@ -52,12 +59,41 @@ func newResourceJSON(r *schema.Resource) resourceJSON {
+			message += fmt.Sprintf("%d x %s\n", count, rType)
+		}
+	}
+	return []string{message}
```
Should we just return a string instead of []string?

sinabakh replied:
Right, will fix it.

sinabakh replied:
The whole function was removed.

aliscott commented on pkg/output/json.go:
```diff
@@ -52,12 +59,41 @@ func newResourceJSON(r *schema.Resource) resourceJSON {
 
-func ToJSON(resources []*schema.Resource) ([]byte, error) {
+func showSkippedResourcesJSON(resources []*schema.Resource, showDetails bool) []string {
+	unSupportedTypeCount, _, unSupportedCount, _ := terraform.CountSkippedResources(resources)
```
"unsupported" is a valid word, so we could use `unsupportedTypeCount` and `unsupportedCount` as the var names. Alternatively we could use `skippedTypeCount` and `skippedCount`.

sinabakh replied:
Changed the words, I used the unsupported because in the future we have free resources which are skipped but are unsupported.

sinabakh replied:
Just changed the "unsupported" case. I think it's better to stick to skipped/unsupported

aliscott commented on pkg/output/table.go:
```diff
@@ -68,6 +93,8 @@ func ToTable(resources []*schema.Resource) ([]byte, error) {
 
 	t.Render()
 
+	showSkippedResourcesTable(resources, c.Bool("show-skipped"), bufw)
```
Could we do something like this:
```go
bufw.WriteString(skippedResourcesMessage(resources, c.Bool("show-skipped")))
```
and put the logic for the message into a shared function in `output/output.go`?

sinabakh replied:
Done

aliscott commented on internal/providers/terraform/provider.go:
```diff
@@ -102,3 +102,22 @@ func generatePlanJSON(dir string, path string) ([]byte, error) {
+	for _, r := range resources {
+		if r.IsSkipped {
+			skippedCount++
+			// FIXME: Count free resources when https://github.com/infracost/infracost/issues/121 is done
```
@sinabakh / @alikhajeh1 do we think we will add support for this any time soon? If yes then it makes sense to leave this in, but if not, I'd be tempted to remove the logic just now.

alikhajeh1 replied:
+1 on removing this FIXME comment line, we don't need the free-resource logic just now.

I see #121 as reviewing/tweaking the existing resources so we support cost components for anything we can easily handle, I guess https://github.com/infracost/infracost/pull/126 is doing that for EC2.

sinabakh replied:
I think we will add the free resources soon, but for now, I will remove the whole free logic and will add it back in its related PR.

sinabakh replied:
Done

aliscott reviewed (commented):
This looks great!

> BREAKING: The json output used to be an array, I changed it to an object.

This is a good idea 👍

alikhajeh1 commented:
If https://github.com/infracost/infracost/pull/126 is merged, we might want to rebase and tweak this branch too or send another PR

sinabakh commented:
@alikhajeh1 I checked the #126 I couldn't understand the need for tweaks? What is changed?
